staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/types.go

```diff
@@ -104,12 +104,23 @@ func (q *queue) Dequeue() (*request, bool) {
 // GetVirtualFinish returns the expected virtual finish time of the request at
 // index J in the queue with estimated finish time G
 func (q *queue) GetVirtualFinish(J int, G float64) float64 {
-	// The virtual finish time of request number J in the queue
-	// (counting from J=1 for the head) is J * G + (virtual start time).
+	var estimatedServiceTime float64
+	currentIdx := 0
+	q.requests.Walk(func(r *request) bool {
+		// starting with the oldest request, we are going
+		// to include request(s) up to index J
+		if currentIdx > J {
+			return false
+		}
+		currentIdx++
 
-	// counting from J=1 for the head (eg: queue.requests[0] -> J=1) - J+1
-	jg := float64(J+1) * float64(G)
-	return jg + q.virtualStart
+		// the estimated service time of a single request is (G * request width)
+		estimatedServiceTime += float64(G) * float64(r.Seats())
+		return true
+	})
+
+	// The virtual finish time of request number J in the queue
+	return q.virtualStart + estimatedServiceTime
```

Comment:
FTR - this doesn't fully fix the linked bug:
https://github.com/kubernetes/apiserver/blob/master/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L746
is still not fixed.

Reply:
actually, I would like to discuss this at today's meeting: here are my thoughts:

`G` = estimated service time = 60s (60 service-seconds)
`S` = actual service time
`W` = seats = width

In the virtual world, the cost (we can compare this to size of a packet) of a request is `P = G*W `
so virtual finish time is `Fi = Si + P`
similarly, when a request completes in the real world, it took `S` seconds, so we adjust the queue's virtual start by subtracting `S` from `P`.

with `(G-S)*W` we have `(GW - SW)` - so let's say `S` = `10s` for two requests, `r1` with `W=1` and `r2` with `W=20`. aren't we being less fair to `r1` (since both requests took the same amount of time in the real world)?

Reply:
Not unfair.  They took the same amount of time, but drew different amounts of power from the server during that time.
BTW, IIRC, the initial guess is no longer a constant.  It's a field in the request, or is it the queue?

Reply:
> similarly, when a request completes in the real world, it took S seconds, so we adjust the queue's virtual start by subtracting S from P.

I don't follow. The S is duration (in seconds) and P is "duration width" (in seconds x meters conceptually). So units don't match.

> with (G-S)*W we have (GW - SW) - so let's say S = 10s for two requests, r1 with W=1 and r2 with W=20. aren't we being less fair to r1 (since both requests took the same amount of time in the real world)?

Why - you're multiply by W.  And we already added G*W previously:
https://github.com/kubernetes/apiserver/blob/master/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go#L640

So here we're just course-correcting.

Reply:
> The S is duration (in seconds) and P is "duration width" (in seconds x meters conceptually). So units don't match.

`W` is an absolute value, so `P=G*W` gives us `P service-seconds` as opposed to `P bit-times`. 
I plugged in `(G-S)*W`, I will go through Mike's enhancement and will write test to experiment in both setups, but this will take time so let's go with `(G-S)*W`.